[DB-14479] In assess-migration payload, send sizing and estimation information

atharva-kurhade-yb opened the pull request:
### Describe the changes in this pull request
Modified callhome and assess migration command to include sizing assessment report information in assess migration payload

Sample Payload JSON :-
{
  "error": "",
  "sizing": {
    "num_nodes": 3,
    "num_sharded_tables": 0,
    "vcpus_per_instance": 4,
    "colocated_reasoning": "Recommended instance type with 4 vCPU and 16 GiB memory could fit 1 objects (1 tables/materialized views and 0 explicit/implicit indexes) with 0.00 MB size and throughput requirement of 0 reads/sec and 0 writes/sec as colocated. Non leaf partition tables/indexes and unsupported tables/indexes were not considered.",
    "memory_per_instance": 16,
    "num_colocated_tables": 1,
    "parallel_voyager_jobs": 1,
    "estimated_time_in_min_for_import": 1,
    "optimial_insert_connections_per_node": 12,
    "optimial_select_connections_per_node": 8
  },
  "iops_interval": 0,
  "schema_summary": "{\"Description\":\"\",\"DbName\":\"\",\"SchemaNames\":null,\"DbVersion\":\"\",\"DatabaseObjects\":[{\"ObjectType\":\"SCHEMA\",\"TotalCount\":1,\"InvalidCount\":0,\"ObjectNames\":\"\"},{\"ObjectType\":\"TABLE\",\"TotalCount\":1,\"InvalidCount\":0,\"ObjectNames\":\"\"}]}",
  "payload_version": "1.0",
  "assessment_issues": "null",
  "target_db_version": "2024.2.0.0",
  "index_sizing_stats": "null",
  "table_sizing_stats": "[{\"object_name\":\"XXXXX\",\"reads_per_second\":0,\"writes_per_second\":0,\"size_in_bytes\":8192}]",
  "source_connectivity": true,
  "migration_complexity": "LOW",
  "migration_complexity_explanation": "Found 0 Level 1 issue(s) and 0 Level 2 issue(s), resulting in LOW migration complexity"
}
### Describe if there are any user-facing changes
<!--
Clarify any changes to the user experience. For instance,
  1. Were there any changes to the command line? 
  2. Were there any changes to the configuration?
  3. Has the installation process changed? 
  4. Were there any changes to the reports? 
-->

### How was this pull request tested?
<!--
Mention if the existing tests were enough
Mention the new unit tests added 
Was any manual testing needed? If yes, is there a need to add integration tests for that?
-->

### Does your PR have changes that can cause upgrade issues? 
| Component        | Breaking changes? |
| :----------------------------------------------: | :-----------: |
| MetaDB                                                    |  Yes/No |
| Name registry json                                        |  Yes/No |
| Data File Descriptor Json                                 |  Yes/No |
| Export Snapshot Status Json                               |  Yes/No |
| Import Data State                                         |  Yes/No |
| Export Status Json                                        |  Yes/No |
| Data .sql files of tables                                 |  Yes/No |
| Export and import data queue                              |  Yes/No |
| Schema Dump                                               |  Yes/No |
| AssessmentDB                                              |  Yes/No |
| Sizing DB                                                 |  Yes/No |
| Migration Assessment Report Json                          |  Yes/No |
| Callhome Json                                             |  Yes (Should not cause any issues as only a new field has been added)|
| YugabyteD Tables                                          |  Yes/No |
| TargetDB Metadata Tables                                  |  Yes/No |

makalaaneesh commented on yb-voyager/cmd/assessMigrationCommand.go:
```diff
@@ -191,9 +191,28 @@ func packAndSendAssessMigrationPayload(status string, errMsg string) {
 
+	sizingRecommendation := callhome.SizingRecommendationCallhome{
+		NumColocatedTables:              len(assessmentReport.Sizing.SizingRecommendation.ColocatedTables),
+		ColocatedReasoning:              assessmentReport.Sizing.SizingRecommendation.ColocatedReasoning,
```
Pls verify that `ColocatedReasoning` and `FailureReasoning` do not include any sensitive information. 
You can look at the code to get an idea of how these fields are set.

atharva-kurhade-yb replied:
ColocatedReasoning does not include sensitive information as it mostly deals with number of colocated and sharded objects and their size and throughput requirements
Example : Recommended instance type with 4 vCPU and 16 GiB memory could fit 1 objects (1 tables/materialized views and 0 explicit/implicit indexes) with 0.00 MB size and throughput requirement of 0 reads/sec and 0 writes/sec as colocated. Non leaf partition tables/indexes and unsupported tables/indexes were not considered.

But Failure Reasoning might contain sensitive information as it catches all the errors that are thrown during sizing assessment.

priyanshi-yb replied:
Let's also bring this up in the standup that we won't be sending the failure reasoning with this change as it includes error msgs. Please also file a task for this, we have to do it at some point after redacting the error msgs.

makalaaneesh commented on yb-voyager/src/callhome/diagnostics.go:
```diff
@@ -99,17 +99,18 @@ type TargetDBDetails struct {
-	IopsInterval                   int64                `json:"iops_interval"`
+	PayloadVersion                 string                          `json:"payload_version"`
+	TargetDBVersion                *ybversion.YBVersion            `json:"target_db_version"`
+	SizingAssessmentReport         *SizingAssessmentReportCallhome `json:"sizing_assessment_report"`
```
perhaps just `sizing` instead of `sizing_assessment_report` ?

atharva-kurhade-yb replied:
Should I change the variable names as well or just the json key?

makalaaneesh reviewed (commented):
In the PR description, pls add a sample json value of the entire payload that gets sent to callhome with your changes.

priyanshi-yb commented on yb-voyager/src/callhome/diagnostics.go:
```diff
@@ -101,6 +101,7 @@ var ASSESS_MIGRATION_CALLHOME_PAYLOAD_VERSION = "1.0"
 type AssessMigrationPhasePayload struct {
 	PayloadVersion                 string               `json:"payload_version"`
 	TargetDBVersion                *ybversion.YBVersion `json:"target_db_version"`
+	Sizing                         *SizingCallhome      `json:"sizing"`
```
As you have added a new field to this structure, tests in `diagnostics_test.go` must fail to test the changes in any structure in this area, mainly to understand whether the change is voyager upgrade safe between the migrations. 
This is just an addition of a new field, so it should be okay, and you can just fix the test.

FYI, please check the table in the PR template whenever you raise a PR it has the check box for the same to update if there is any change to any struct of the voyager state.

atharva-kurhade-yb replied:
In the PR template I would have to set CallhomeJSON to Yes right?

priyanshi-yb replied:
Yes you can, but mention that this is just an additional field so shouldn't affect anything.

atharva-kurhade-yb replied:
Ok thank you for the review.

priyanshi-yb commented on yb-voyager/src/callhome/diagnostics.go:
```diff
@@ -121,6 +122,19 @@ type AssessmentIssueCallhome struct {
+	NumShardedTables                int     `json:"num_sharded_tables"`
+	NumNodes                        float64 `json:"num_nodes"`
+	VCPUsPerInstance                int     `json:"vcpus_per_instance"`
+	MemoryPerInstance               int     `json:"mem_per_instance"`
```
keep the JSON tag the same as `memory_per_instance`.

atharva-kurhade-yb commented:
@priyanshi-yb 
I have added changes to diagnostics_test.go and also changed the field name to memory_per_instance. 
Please review the changes when you get the time

priyanshi-yb reviewed (approved):
LGTM

atharva-kurhade-yb commented:
Also ran a jenkins pipeline to run Oracle tests:-
https://jenkins.dev.yugabyte.com/job/users/job/yb-voyager-testing/job/yb-voyager-testing-pipeline/4651/

sanyamsinghal commented on yb-voyager/cmd/assessMigrationCommand.go:
```diff
@@ -184,9 +184,23 @@ func packAndSendAssessMigrationPayload(status string, errMsg string) {
 	}
 
+	sizingAssessmentReport := callhome.SizingCallhome{
+		NumColocatedTables:              len(assessmentReport.Sizing.SizingRecommendation.ColocatedTables),
```
nit: @atharva-kurhade-yb you can define a varaible which reference to the field `assessmentReport.Sizing.SizingRecommendation` of `assessmentReport` struct and use here to populate callhome.SizingCallhome struct
